command-line-interface.md

```diff
@@ -0,0 +1,150 @@
+# Operations
+
+A conformant [runtime][] MUST provide an executable (called `funC` in the following examples).
+That executable MUST support commands with the following template:
+
+```sh
+$ funC [global-options] <COMMAND> [command-specific-options] <command-specific-arguments>
+```
+
+## Global options
+
+None are required, but the runtime MAY support options that start with at least one hyphen.
+Global options MAY take positional arguments (e.g. `--log-level debug`).
+Command names MUST NOT start with hyphens.
+The option parsing MUST be such that `funC <COMMAND>` is unambiguously an invocation of `<COMMAND>` (even for commands not specified in this document).
+If the runtime is invoked with an unrecognized command, it MUST exit with a nonzero exit code and MAY log a warning to stderr.
+
+## Character encodings
+
+This API specification does not cover character encodings, but runtimes SHOULD conform to their native operating system.
+For example, POSIX systems define [`LANG` and related environment variables][posix-lang] for [declaring][posix-locale-encoding] [locale-specific character encodings][posix-encoding], so a runtime in an `en_US.UTF-8` locale SHOULD write its [version](#version) to stdout in [UTF-8][].
+
+## Commands
+
+### version
+
+Print the runtime version and exit.
+
+* *Options* None are required, but the runtime MAY support options.
+* *Standard streams*
+    * *stdin:* The runtime MUST NOT attempt to read from its stdin.
+    * *stdout:* The runtime MUST print its name, a space, and its version as the first line to its stdout.
+      The name MAY contain any Unicode characters, but MUST NOT contain control codes or newlines.
+      The runtime MAY print additional lines to its stdout, and the format for those lines is not specified in this document.
+    * *stderr:* The runtime MAY print diagnostic messages to stderr, and the format for those lines is not specified in this document.
+* *Exit code:* The runtime MUST exit with zero.
+
+#### Example
+
+```
+$ funC version
+funC 1.0.0
+Built for x86_64-pc-linux-gnu
+$ echo $?
+0
+```
+
+### start
+
+[Start][start] a container from a [bundle directory][bundle].
+
+* *Options*
+    * *`--id <ID>`* Set the container ID when creating a container.
```

Comment:
On Fri, Jul 08, 2016 at 09:14:15PM -0700, Aleksa Sarai wrote:
> > +[Start][start] a container from a [bundle directory][bundle].
> > +
> > +* *Options*
> > +    * *`--id <ID>`* Set the container ID when creating a container.
>
> How is this meant to interact with `funC state` and similar
> commands, which require an ID? IMO you can't have it both ways (you
> need to know the ID at all times).

You don't need to specify ID if: you know you'll never need to call
‘state’, ‘start’, ‘delete’, ….  That was the case before the
create/start split, and I don't see why cases that just want to launch
a container in a namespace without hooks, etc. can't keep that
functionality.  Although as I said above [1], this depends on how #508
plays.

[1]: https://github.com/opencontainers/runtime-spec/pull/513#discussion_r70109722